File: Data Structures & Algorithms/trapping-rain-water/submission-0.java
public class Solution {
    public int trap(int[] height) {
        if (height == null || height.length == 0) {
            return 0;
        }

        int l = 0;
        int r = height.length - 1;
        int leftMax = height[l];   // 2, 
        int rightMax = height[r];  // 1, 2, 3

        int res = 0;
        while (l < r) {
            if (leftMax < rightMax) {
                l++;
                leftMax = Math.max(leftMax, height[l]);
                res += leftMax - height[l];
            } else {
                r--;
                rightMax = Math.max(rightMax, height[r]);
                res += rightMax - height[r];
            }
        }
        return res;
    }
}

File: Data Structures & Algorithms/eating-bananas/submission-0.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        # h = hours to eat
        # piles[i] = # of banaas in that pile
        # k = bananas per hour eating rate - We decide
        # If that pile less that k banas, we may finish this pile but wait until next
        # Find min k to eat banas within hours

        l = 1
        r = max(piles)

        res = r

        while l <= r:
            k = (l+r) // 2
            totalTime = 0
            for p in piles:
                totalTime += math.ceil(float(p) / k)

            print(totalTime)
            if totalTime <= h:
                res = k
                r = k - 1
            else:
                l = k + 1

        return res